### PR TITLE
Reverter

### PR DESCRIPTION
**Pull request by jdkasten**
Here are the changes...

1. Split up augeas_configurator.py into augeas_configurator.py and reverter.py
reverter.py contains all of the configuration logging and rollback features that should be usable by more classes than just the Augeas-based configurators.  

2. Improved the reverter class and added 100% test coverage of Reverter class.  I think the testing is pretty thorough.
Having a separate Reverter class allowed...

3.  Better main.py startup functionality.  I have started to place more appropriate errors in the Installers.  They should be able to fail with NoInstallation (if for instance Apache can't be found), or Misconfigured.  The startup scripts can use the new minimalist Reverter class to view config changes and to rollback the configuration when the webserver is misconfigured.
Side note about Misconfigured initialization error:
The ApacheConfigurator can't fully initialize without running a module test that fails if the server is misconfigured.  The module information is necessary in order to accurately parse the config files. I imagine this problem of initialization misconfiguration will generalize.

4.  The Revoker class doesn't need a functioning installer to act appropriately.  It can get much of the functionality without the installer.  So I made it optional.

5.  I attempted to make sure that client.py was safe from having partial functionality. (None Installers/Authenticators)

6. Caught a few bugs or potential bugs along the way.  If you have noticed your /var/lib/letsencrypt folder has filled with nonce.crt files, I found the source of that problem in dvsni.py and made sure it couldn't happen again.

I believe everything works, pylint, pep8 ect... but I will go over formatting/documentation again.
Input is appreciated.

### CODE REVIEW

File: docs/api/client/reverter.rst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+:mod:`letsencrypt.client.reverter`
+---------------------------------
```
Comment (kuba) [inline docs/api/client/reverter.rst]: Be careful, there are not enough `-`... You have to "underline" all characters in the headline, otherwise you'll get `WARNING: Title underline too short` (also c.f. 48f1497af6a5fd4c787013eb28a67398680aff74)

File: letsencrypt/client/crypto_util.py [hunk 2]
```diff
@@ -45,20 +45,20 @@ def create_sig(msg, key_str, nonce=None, nonce_len=CONFIG.NONCE_SIZE):
...
-        "nonce": le_util.jose_b64encode(nonce),
-        "alg": "RS256",
-        "jwk": {
-            "kty": "RSA",
-            "n": le_util.jose_b64encode(n_bytes),
-            "e": le_util.jose_b64encode(e_bytes),
+        'nonce': le_util.jose_b64encode(nonce),
+        'alg': "RS256",
+        'jwk': {
+            'kty': "RSA",
```
Comment (kuba) [inline letsencrypt/client/crypto_util.py]: "Be consistent with your choice of string quote character within a file. " (https://google-styleguide.googlecode.com/svn/trunk/pyguide.html?showone=Strings#Strings)

File: letsencrypt/client/apache/configurator.py [hunk 3]
```diff
@@ -523,6 +519,7 @@ def make_vhost_ssl(self, nonssl_vhost):
 
         return ssl_vhost
 
+    # pylint: disable=no-method-argument,no-self-use,unused-argument
```
Comment (kuba) [inline letsencrypt/client/apache/configurator.py]: what's going on here? also, this probably removes pylint checking not only for the following method, but for the entire code block indented at this level...

File: letsencrypt/client/display.py [hunk 4]
```diff
@@ -74,12 +75,20 @@ def success_installation(self, domains):
...
+            if c['installed']:
+                if c['installed'] == CONFIG.CERT_DELETE_MSG:
+                    status = "Deleted/Moved"
+                else:
+                    status = "Installed"
+            else:
+                status = ""
+
+            list_choices.append((str(i+1), "{0} | {1} | {2}".format(
+                str(c['cn'].ljust(self.width-39)),
```
Comment (kuba) [inline letsencrypt/client/display.py]: style: space around operator

File: letsencrypt/client/display.py [hunk 5]
```diff
@@ -74,12 +75,20 @@ def success_installation(self, domains):
...
+                c['not_before'].strftime("%m-%d-%y"),
+                status)))
```
Comment (kuba) [inline letsencrypt/client/display.py]: Could you, please, provide unittest for this method?

File: letsencrypt/client/CONFIG.py [hunk 6]
```diff
@@ -89,23 +90,7 @@
 spdy, TODO
 
 """
-
-# ENHANCEMENTS = [
```
Comment (kuba) [inline letsencrypt/client/CONFIG.py]: :+1: for removing commented code!

File: letsencrypt/scripts/main.py [hunk 7]
```diff
@@ -201,27 +215,95 @@ def read_file(filename):
         raise argparse.ArgumentTypeError(exc.strerror)
 
 
-def rollback(installer, checkpoints):
+def rollback(checkpoints):
```
Comment (kuba) [inline letsencrypt/scripts/main.py]: Could we please move `rollback` and other bigger chunks of code to `letsencrypt.client` and keep `scripts` as short as possible?

File: letsencrypt/client/tests/reverter_test.py [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+        # Now Run tests
+        # These were new files.. they should be removed
+        self.assertFalse(os.path.isfile(config3))
+        self.assertFalse(os.path.isfile(config4))
+
+        # Check to make sure everything got rolled back appropriately
+        self.assertEqual(read_in(self.config1), "directive-dir1")
+        self.assertEqual(read_in(self.config2), "directive-dir2")
+
+# pylint: disable=invalid-name,protected-access,too-many-instance-attributes
```
Comment (kuba) [inline letsencrypt/client/tests/reverter_test.py]: Again, please correct me if I'm wrong, but IMHO this affects all code indented at this level. Also, it's not clear which lines cause the `invalid-name` error (I would put it on the line of offending definition to make things clear). Similarly for `protected-access`, but... if it's protected, then maybe some other code should be testing it?

File: letsencrypt/client/tests/reverter_test.py [hunk 9]
```diff
@@ -0,0 +1,459 @@
...
+import logging
+import os
+import shutil
+import tempfile
+import unittest
+
+import mock
+
+
+# pylint: disable=invalid-name,protected-access,too-many-instance-attributes
```
Comment (kuba) [inline letsencrypt/client/tests/reverter_test.py]: https://github.com/letsencrypt/lets-encrypt-preview/pull/185/files#r23503203

File: letsencrypt/client/reverter.py [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+                protected_files.extend(protected_fd.read().splitlines())
+
+        # Verify no save_file is in protected_files
+        for filename in protected_files:
+            if filename in save_files:
+                raise errors.LetsEncryptReverterError(
+                    "Attempting to overwrite challenge "
+                    "file - %s" % filename)
+
+    # pylint: disable=no-self-use, anomalous-backslash-in-string
```
Comment (kuba) [inline letsencrypt/client/reverter.py]: https://github.com/letsencrypt/lets-encrypt-preview/pull/185/files#r23503203

`anomalous-backslash-in-string` is not necessary if `r"""` is used.

File: letsencrypt/client/reverter.py [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+            except errors.LetsEncryptReverterError:
+                # We have a partial or incomplete recovery
+                logging.fatal("Incomplete or failed recovery for IN_PROGRESS "
+                              "checkpoint - %s",
+                              self.direc['progress'])
+                raise errors.LetsEncryptReverterError(
+                    "Incomplete or failed recovery for IN_PROGRESS checkpoint "
+                    "- %s" % self.direc['progress'])
+
+    # pylint: disable=no-self-use
```
Comment (kuba) [inline letsencrypt/client/reverter.py]: https://github.com/letsencrypt/lets-encrypt-preview/pull/185/files#r23503203

File: letsencrypt/client/reverter.py [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+        except (IOError, OSError):
+            logging.fatal(
+                "Unable to remove filepaths contained within %s", file_list)
+            raise errors.LetsEncryptReverterError(
+                "Unable to remove filepaths contained within "
+                "{0}".format(file_list))
+
+        return True
+
+    # pylint: disable=no-self-use
```
Comment (kuba) [inline letsencrypt/client/reverter.py]: https://github.com/letsencrypt/lets-encrypt-preview/pull/185/files#r23503203

File: letsencrypt/client/tests/reverter_test.py (hunk 8 above)
Comment (jdkasten) [inline letsencrypt/client/tests/reverter_test.py]: This goes back to my complaint in #178.  I think I had about 7 or 8 invalid-names.  The protected access was for testing the private functions within the Reverter class.  I think it is reasonable to test them from within reverter_test.py.  I don't see another alternative... I agree that this should have probably gone on offending line... but I was pretty fed up at pylint by this point.  Again this goes back to my argument in #178, but I will fix this.

File: letsencrypt/client/revoker.py [hunk 13]
```diff
@@ -30,16 +30,16 @@ def acme_revocation(self, cert):
         :rtype: dict
 
         """
```
Comment (jdkasten) [inline letsencrypt/client/revoker.py]: this has all been reset...

File: letsencrypt/client/tests/reverter_test.py (hunk 8 above)
Comment (kuba) [inline letsencrypt/client/tests/reverter_test.py]: yup, i was right about how pylint treats locally disabled messages for block, c.f. #189

File: letsencrypt/client/display.py (hunk 5 above)
Comment (jdkasten) [inline letsencrypt/client/display.py]: I decided to completely remove this code from the PR.  It didn't actually implement what I believe to be necessary (it didn't go far enough).  I think the revoker/display class need overhauls, but I decided to leave it to another PR.

**Comment by jdkasten** [comment]
I believe everything has been addressed.  

Pylint disable messages have been more appropriately handled.
Many of the big functions in main.py have been moved to client.py.  
The rollback function has been refactored for readability.

I have removed all of the new revoker code (I did add a temporary guard to make sure the installer existed).  My attempt only addressed half of the problems and a more complete solution should come in a follow-up PR.

File: letsencrypt/client/tests/client_test.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+        self._call(1)
+
+        self.assertEqual(self.m_install().rollback_checkpoints.call_count, 1)
+        self.assertEqual(self.m_install().restart.call_count, 1)
+
+    @mock.patch("letsencrypt.client.reverter.Reverter")
+    @mock.patch("letsencrypt.client.client.determine_installer")
+    def test_misconfiguration_fixed(self, mock_det, mock_rev):
+        from letsencrypt.client.errors import LetsEncryptMisconfigurationError
```
Comment (kuba) [inline letsencrypt/client/tests/client_test.py]: This should be imported at the top as `from letsencrypt.client import errors`. I can see there are a lot of similar issues, so I will let you hunt them down.

File: letsencrypt/client/tests/client_test.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+import mock
+import zope.component
+
+
+class RollbackTest(unittest.TestCase):
+    """Test the rollback function."""
+    def setUp(self):
+        self.m_install = mock.MagicMock()
+        self.m_input = mock.MagicMock()
+        zope.component.getUtility = self.m_input
```
Comment (kuba) [inline letsencrypt/client/tests/client_test.py]: this looks dangerous, as `zope.component.getUtility` never gets restored, patching should be used

File: letsencrypt/client/tests/client_test.py [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+
+    def _call(self, checkpoints):  # pylint: disable=no-self-use
```
Comment (kuba) [inline letsencrypt/client/tests/client_test.py]: To stay consistent with my recent changes at 23cab4e6943ff1a61f985dd67d2ebcf115ca9775 instead of `pylint: disable=no-self-use` you should use

```
@classmethod
def _call(cls, checkpoints):
    ...
```

Here, and in other `_call`s.

File: letsencrypt/client/tests/reverter_test.py [hunk 17]
```diff
@@ -0,0 +1,467 @@
...
+
+
+class ReverterCheckpointLocalTest(unittest.TestCase):
+    # pylint: disable=too-many-instance-attributes
+    """Test the Reverter Class."""
+    def setUp(self):
+        from letsencrypt.client.reverter import Reverter
+
+        # Disable spurious errors... we are trying to test for them
+        logging.disable(logging.CRITICAL)
```
Comment (kuba) [inline letsencrypt/client/tests/reverter_test.py]: `tearDown` should probably revert `logging` to its initial state.

File: letsencrypt/scripts/main.py [hunk 18]
```diff
@@ -77,29 +76,36 @@ def main():  # pylint: disable=too-many-statements
...
         display_eula()
 
+    # Make sure we actually get an installer that is functioning properly
+    # before we begin to try to use it.
+    try:
+        installer = client.determine_installer()
+    except errors.LetsEncryptMisconfigurationError as err:
+        logging.fatal("Please fix your configuration before proceeding.  "
+                      "The Installer exited with the following message: "
+                      "%s", str(err))
```
Comment (kuba) [inline letsencrypt/scripts/main.py]: spurious cast to `str`, `"%s"` get the job done for you

File: letsencrypt/scripts/main.py [hunk 19]
```diff
@@ -12,12 +12,11 @@
 from letsencrypt.client import client
 from letsencrypt.client import display
 from letsencrypt.client import interfaces
+from letsencrypt.client import errors
```
Comment (kuba) [inline letsencrypt/scripts/main.py]: this should go one line higher

**Comment by kuba** [comment]
You can bump up coverage to 66%

**Comment by jdkasten** [comment]
Thanks @Kuba!  I was knocking them off as I was receiving emails.  I believe i have addressed all comments up to the coverage bump.

File: letsencrypt/client/tests/reverter_test.py [hunk 20]
```diff
@@ -0,0 +1,451 @@
...
+
+        tup = setup_test_files()
+        self.config1, self.config2, self.dir1, self.dir2, self.sets = tup
+
+    def tearDown(self):
+        shutil.rmtree(self.work_dir)
+        shutil.rmtree(self.dir1)
+        shutil.rmtree(self.dir2)
+
+        logging.disable(logging.NOTSET)
```
Comment (kuba) [inline letsencrypt/client/tests/reverter_test.py]: Same as in https://github.com/letsencrypt/lets-encrypt-preview/pull/185/files#r23518302

File: letsencrypt/client/tests/reverter_test.py [hunk 21]
```diff
@@ -0,0 +1,451 @@
...
+        self.assertEqual(read_in(self.config2), "directive-dir2")
+
+
+class TestFullCheckpointsReverter(unittest.TestCase):
+    # pylint: disable=too-many-instance-attributes
+    """Tests functions having to deal with full checkpoints."""
+    def setUp(self):
+        from letsencrypt.client.reverter import Reverter
+        # Disable spurious errors...
+        logging.disable(logging.CRITICAL)
```
Comment (kuba) [inline letsencrypt/client/tests/reverter_test.py]: Same as in https://github.com/letsencrypt/lets-encrypt-preview/pull/185/files#r23518302

File: letsencrypt/client/reverter.py [hunk 22]
```diff
@@ -0,0 +1,434 @@
...
+        # There should only be timestamped (float) directories
+        try:
+            for bkup in backups:
+                float(bkup)
+        except ValueError:
+            raise errors.LetsEncryptReverterError(
+                "Invalid directories in {0}".format(self.direc['backup']))
+
+        for bkup in backups:
+            print time.ctime(float(bkup))
```
Comment (kuba) [inline letsencrypt/client/reverter.py]: aren't these `print` statements relicts of debugging? they would break down ncurses, etc.

Comment (jdkasten) [inline letsencrypt/client/reverter.py]: I agree that this isn't good long-term.  At one point I did refactor them to be logging.info statements... but they don't appear on the screen too well.  I figured we needed a display method that wrapped the output and allowed you to scroll.

Perhaps the display.generic_notification window does do this...  I should have put a TODO... but I kind of wanted to look at/document/rearrange the display code before I spent time modifying this code.

I will do some quick experiments with generic_notification... might as well fix this now.

Comment (jdkasten) [inline letsencrypt/client/reverter.py]: As I feared... it wasn't a simple fix.  I had to modify the display class functions...and the API... since I was editing the display API... I went ahead and documented all of the functions while I was at it...just to keep everything consistent.  The class still needs some work... but at least the functions are documented and the notifications work.

view_config_changes now correctly uses the display output of choice.  I am not 100% satisfied with the FileDisplay generic_notification especially... but it is a strict improvement on what is there.

I did do a documentation check to make sure that SPHINX ran without warnings.

**Comment by jdkasten** [comment]
This is not ready to go. I have realized that I forgot to clean up client_test.py

**Comment by kuba** [comment]
Thanks a lot, especially for improving docs over there.

**Comment by jdkasten** [comment]
I believe this is ready now.  It does fix #166 but leaves some related problems unfinished.  Those will come in follow-up PRs.

**Comment by kuba** [comment]
Good work :+1:
